### PR TITLE
Checks shape of boxes and coords for potentials

### PR DESCRIPTION
**Pull request by badisa**
* Previously could provide incorrectly shaped objects and still have potentials run

### CODE REVIEW

**Comment by badisa** [comment]
- [ ] Add verification that boxes are ortholinear
- [ ] Remove ortholinear check in Nonbonded potential

**Review by maxentile — APPROVED** [review]
Looks good -- should improve safety!

(Partially addresses https://github.com/proteneer/timemachine/issues/562 -- harder to check the `num_atoms` dimension of coords, or to check leading dimension of parameter arrays, possibly something to revisit in the future.)

File: tests/test_potentials.py [hunk 1]
```diff
@@ -37,6 +37,41 @@ def test_bound_potential_get_potential(harmonic_bond):
     assert unbound_impl is bound_impl.get_potential()
 
 
+def test_bound_potential_execute_validation(harmonic_bond):
+    bound_impl = harmonic_bond.bound_impl(np.float32)
+
+    with pytest.raises(RuntimeError, match="coords must have at least 2 dimensions"):
```
[the remark this answers is not in the document]
Comment (badisa) [reply tests/test_potentials.py]: Strengthened in https://github.com/proteneer/timemachine/pull/916/commits/25e0f14d39e65109ab26e06418f50abebfc62e3b

File: tests/test_potentials.py [hunk 2]
```diff
@@ -37,6 +37,57 @@ def test_bound_potential_get_potential(harmonic_bond):
...
+        with pytest.raises(RuntimeError, match="box must be 3x3"):
+            execute_method(np.zeros((1, 3)), harmonic_bond.params, np.ones((2, 2)), *extra_params)
+
+        with pytest.raises(RuntimeError, match="box must be ortholinear"):
+            execute_method(np.zeros((1, 3)), harmonic_bond.params, np.ones((3, 3)), *extra_params)
+
+        with pytest.raises(RuntimeError, match="box must have positive values along diagonal"):
+            execute_method(np.zeros((1, 3)), harmonic_bond.params, np.eye(3) * 0.0, *extra_params)
+
+        execute_method(np.zeros((3, 3)), harmonic_bond.params, np.eye(3), *extra_params)
```
Comment (mcwitt) [inline tests/test_potentials.py:88]: nit: is it possible to DRY these with the analogous tests for bound potentials above?

File: timemachine/cpp/src/wrap_kernels.cpp [hunk 3]
```diff
@@ -30,6 +30,35 @@
...
+    if (coords.shape(coord_dimensions - 1) != 3) {
+        throw std::runtime_error("coords must have a shape that is 3 dimensional");
+    }
+    if (box.ndim() != 2) {
+        throw std::runtime_error("box dimensions must be 2");
+    }
+
+    if (box.shape(0) != 3 || box.shape(1) != 3) {
+        throw std::runtime_error("box must be 3x3");
+    }
```
Comment (mcwitt) [inline timemachine/cpp/src/wrap_kernels.cpp:49]: nit: could these be combined like
```suggestion
    if (box.ndim() != 2 || box.shape(0) != 3 || box.shape(1) != 3) {
        throw std::runtime_error("box must be 3x3");
    }
```

File: timemachine/cpp/src/wrap_kernels.cpp [hunk 4]
```diff
@@ -30,6 +30,35 @@
...
+    auto box_data = box.data();
+    for (int i = 0; i < box.size(); i++) {
```
Comment (mcwitt) [inline timemachine/cpp/src/wrap_kernels.cpp:51]: nit: any reason not to iterate only over the diagonal here as is done for the box width checks above?

**Review by mcwitt — APPROVED** [review]

File: tests/test_potentials.py (hunk 2 above)
Comment (badisa) [reply tests/test_potentials.py:79]: Done in https://github.com/proteneer/timemachine/pull/916/commits/3566d1dbed13e8583cefd91846bdd729a3a0721f

**Pull request merged**